exec/call.go

```diff
@@ -0,0 +1,73 @@
+// Copyright 2017 The go-interpreter Authors.  All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package exec
+
+import "errors"
+
+func (vm *VM) doCall(compiled compiledFunction, index int64) {
+	newStack := make([]uint64, compiled.maxDepth)
+	locals := make([]uint64, compiled.totalLocalVars)
+
+	for i := compiled.args - 1; i > -1; i-- {
+		locals[i] = vm.popUint64()
+	}
+
+	// save execution context
+	prevCtxt := vm.context
+
+	vm.context = &context{
+		stack:    newStack,
+		stackTop: -1,
+		locals:   locals,
+		code:     compiled.code,
+		pc:       0,
+		curFunc:  index,
+	}
+
+	rtrn := vm.execCode(compiled)
+
+	// restore execution context
+	vm.context = prevCtxt
+
+	if compiled.returns {
+		vm.pushUint64(rtrn)
+	}
+}
+
+var ErrSignatureMismatch = errors.New("exec: signature mismatch in call_indirect")
+
+func (vm *VM) call() {
+	index := vm.fetchUint32()
+	vm.doCall(vm.compiledFuncs[index], int64(index))
+}
+
+func (vm *VM) callIndirect() {
+	index := vm.fetchUint32()
+	fnExpect := vm.module.Types.Entries[index]
+	_ = vm.fetchUint32() // reserved (https://github.com/WebAssembly/design/blob/27ac254c854994103c24834a994be16f74f54186/BinaryEncoding.md#call-operators-described-here)
+	index = vm.popUint32()
+	fnActual := vm.module.FunctionIndexSpace[vm.module.TableIndexSpace[0][index]]
+
+	if len(fnExpect.ParamTypes) != len(fnActual.Sig.ParamTypes) {
+		panic(ErrSignatureMismatch)
```

Comment:
The WebAssembly spec says that implementations should "trap" in case of such errors in runtime: https://github.com/WebAssembly/design/blob/27ac254c854994103c24834a994be16f74f54186/Semantics.md#traps